IO mapping as per @ledvinap and adjustments to i2c driver

blckmn opened the pull request:
Updating i2c to use mapping, and bringing in the pull request https://github.com/cleanflight/cleanflight/pull/1421 as raised by @ledvinap for IO mapping (and allowing for F4 targets)

Not all targets good yet - will continue to test and update this pull request.

blckmn commented:
I've tested this on a Naze and the Bluejay. Just need to fix up the other targets. Is it possible @rs2k you can test on Revo, and @MJ666 on the Alien?

ledvinap commented on src/main/drivers/bus_i2c.c:
```diff
@@ -0,0 +1,441 @@
+#endif
+
+static i2cDevice_t i2cHardwareMap[] = {
+    { I2C1, IO_TAG(I2C1_SCL), IO_TAG(I2C1_SDA), RCC_APB1(I2C1), false, I2C1_EV_IRQn, I2C1_ER_IRQn },
```
designated initializer should be used here ... 
Sorry, the I2C code in my repository was just quick rewrite to get rid of old IO code, it needs some polishing ...
I'll try to get to it soon

blckmn replied:
Haven't got to the passing in of configuration yet. The idea was to configure mag, mpu etc to be attached to an i2c (or spi) instance (by index into that AF). And then configure the AF to use certain pins.

ledvinap replied:
So that you can use single I2C on different pins concurrently? Will work, but sounds a bit too complicated.

Or select one configuration on runtime? Just add array of possible pinTag+AF configurations (limited to some reasonable number on F3).

If fully general IO configuration is supported, then it would be probably best to tag IO pin with function:
`PA3 - I2C(1)` and then search possible configurations to get correct AF (this is important on F3 ADC - single pin is connected to different ADCs, it is not easy to select 'best' one)

blckmn replied:
Agreed there will be a limited combination of pins applicable for hardware i2c but soft_i2c will mean any pin could be used. The device (mag, mpu, baro etc) should not know whether it is soft or hard i2c they are just attached to an i2c instance.

In the hardware driver config - you'd presumably have a lookup that would "warn" / "disable" the function if the pin combination was not supported.

Initially my focus is to get the infrastructure working - i.e. i2cRead / Write passing instance etc - and then clean up the configuration of the AF underneath the interface. 

This is only i2c. SPI will be even harder as 3 pins (plus CS) will make it a bigger possible combo set - hence initially no checks on whether a pin is valid or not.

ledvinap commented on src/main/drivers/bus_i2c.c:
```diff
@@ -0,0 +1,441 @@
+
+void i2cInit(I2CDevice index) 
+{
+    i2cDevice_t device;
```
Mmixing `device` as index and `device` as pointer to device descriptor is bad. Try to use consistent naming

blckmn replied:
That's because I've been flip flopping ... I'll clean this up.

MJ666 commented:
Tested on my AlienFlight F4 development platform and everything looks to be normal. I2C sensor is working and Led's also. Our prototypes don't use I2C yet.

blckmn commented:
Awesome. Thx @MJ666

blckmn committed:
Initial IO mapping

Remove processor specific files for led, sound.
Combine i2c for both F1 and F4 targets
Add QUANTON target
Add LUX_RACE target
Update F4 targets to use IO in serial_uart_stm32f4xx

blckmn commented:
Hi @mj666 yes - I'm hoping to get this merged ASAP - so we can then merge in your pr - and adjust to take into account your improvements.

Happy to ditch the directories... :)

blckmn committed:
Move SPI to use IO mapping and hardwareMap

rs2k commented:
Sorry for the delay. I've been traveling. I'll get this merged today if it's ready. :)

blckmn commented:
It's tested on bluejay, alien and naze... There will be some broken older F1 targets - im happy to work with whomever to fix those as they are discovered.

So good to merge. 

I think I'm going to have to buy a few more different targets for testing.

blckmn commented:
I wonder if we should ditch some of the older targets... and focus on F3 and F4?

rs2k commented:
We probably should. I'd like to keep the CC3D going a bit longer at least assuming it doesn't take too much time away.

blckmn commented:
Yes - CC3D and NAZE (but Rev 5 and 6 only)

ledvinap commented:
This code may actually decrease image size a bit after all old GPIO code is replaced ... but until than the code is duplicated

rs2k commented:
It looks like there are some conflicts in some target.h files, possibly from https://github.com/rs2k/raceflight/pull/75. @blckmn, what do you think?

blckmn commented:
I'll rebase tonight and fix them up :)

blckmn commented:
All re-based and good to go @rs2k

rs2k commented:
Merged in and seems good. I created some test firmware for the targets that compiled. The VRCore and none of the F3s or F1s compile.

blckmn commented:
I'll check the F1/3 tonight and add a new pull request to fix them up. Any F1 targets you want to flick let me know, I'll drop them in the same pull request.

ledvinap commented:
Ping me if I can help ...

blckmn commented:
Thanks @ledvinap I'll let you know. Should be easy enough fixes...
